src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuilding.java

```diff
@@ -1337,159 +1336,341 @@ public void setMirror()
+        getOpenRequests(data).forEach(request ->
+        {
+            getColony().getRequestManager().updateRequestState(request.getToken(), RequestState.CANCELLED);
+
+            openRequests.get(request.getRequest().getClass()).remove(request.getToken());
+            if (openRequests.get(request.getRequest().getClass()).isEmpty())
+            {
+                openRequests.remove(request.getRequest().getClass());
+            }
+
+            requestsByCitizen.remove(request.getToken());
+        });
+
+        getCompletedRequests(data).forEach(request -> getColony().getRequestManager().updateRequestState(request.getToken(), RequestState.RECEIVED));
+
+        if (citizensByRequests.containsKey(data.getId()))
         {
-            return buildingMaxLevel;
+            citizensByRequests.remove(data.getId());
         }
 
-        /**
-         * Checks if this building is at its max level.
-         *
-         * @return true if the building is at its max level.
-         */
-        public boolean isBuildingMaxLevel()
+        if (citizensByCompletedRequests.containsKey(data.getId()))
         {
-            return buildingLevel >= buildingMaxLevel;
+            citizensByCompletedRequests.remove(data.getId());
         }
 
-        /**
-         * Get the current work order level.
-         *
-         * @return 0 if none, othewise the current level worked on
-         */
-        public int getCurrentWorkOrderLevel()
+        markDirty();
+    }
+
+    /**
+     * Overrule the next open request with a give stack.
+     * <p>
+     * We squid:s135 which takes care that there are not too many continue statements in a loop since it makes sense here
+     * out of performance reasons.
+     *
+     * @param stack the stack.
+     */
+    @SuppressWarnings("squid:S135")
+    public void overruleNextOpenRequestWithStack(@NotNull final ItemStack stack)
+    {
+        if (ItemStackUtils.isEmpty(stack))
         {
-            return workOrderLevel;
+            return;
         }
 
-        /**
-         * Get the current work order level.
-         *
-         * @return 0 if none, othewise the current level worked on
-         */
-        public boolean hasWorkOrder()
+        for (int citizenId : citizensByRequests.keySet())
+        {
+            final CitizenData data = getColony().getCitizen(citizenId);
+
+            if (data == null)
+            {
+                continue;
+            }
+
+            final IRequest<? extends IDeliverable> target = getOpenRequestsOfTypeFiltered(data, TypeToken.of(IDeliverable.class),
+              request -> request.getRequest().matches(stack)).stream().findFirst().orElse(null);
+
+            if (target == null)
+            {
+                continue;
+            }
+
+            getColony().getRequestManager().overruleRequest(target.getToken(), stack.copy());
+            return;
+        }
+    }
+
+    public <R> ImmutableList<IRequest<? extends R>> getOpenRequestsOfTypeFiltered(
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Remove usage of generic wildcard type. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orionminecraft.com/coding_rules#rule_key=squid%3AS1452)